### PR TITLE
Percent-encode grpc-message

### PR DESCRIPTION
**Pull request by note**
## Problem

Having such code:

```
new GrpcServiceException(Status.INVALID_ARGUMENT.withDescriptionSafe(e.getMessage))
```

may result in issuing invalid response if `e.getMessage` contains `\n` character. Since `grpc-message` is being transported with http headers it cannot contain arbitrary characters, see more [here](https://github.com/grpc/grpc-go/issues/576). What's recommended by standard is to percent-encode the message, see the [spec](https://github.com/grpc/grpc/issues/4672).

It's been added to grpc-java in [this PR](https://github.com/grpc/grpc-java/pull/1517/files).

## Why it is relevant

I saw this issue happening in a production setting. scalapb error messages can contain `\n` and it's getting problematic if you use linkerd. Linkerd will treat such response as incorrect. As a consequence the client will not see the original `grpc-status: 3, grpc-message: meaningful message` but `grpc-status: 13, grpc-message: protocol error: unspecific protocol error detected`.

## Solution

Copy and adapt grpc-java percent-encoding implementation.

## Other

I am not sure what's the most appropriate way of giving credits to grpc-java. Both project are on Apache 2.0 license, not sure what's required from license standpoint

### CODE REVIEW

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 1]
```diff
@@ -81,19 +87,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
   override def renderInResponses = true
   override val companion = `Status-Message`
+  override def value() = toAsciiString(unencodedValue)
+
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L582
+  private def toAsciiString(value: String): String = {
+    val valueBytes = value.getBytes(StandardCharsets.UTF_8)
+    val firstIndexToEscape = valueBytes.indexWhere(isEscapingChar(_))
+    if (firstIndexToEscape == -1)
+      new String(valueBytes)
```
Comment (lomigmegard) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala:104]: return `value` instead of re-wrapping inside a new String?

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 2]
```diff
@@ -81,19 +87,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
 
   def findIn(headers: immutable.Seq[HttpHeader]): Option[String] =
     headers.collectFirst { case h if h.is(name) => h.value() }
+
+  private val HexArr: Array[Byte] =
+    Array('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L626
+  private def parseAsciiString(value: String): String = {
+    val transferEncoding = Charset.forName("US-ASCII")
```
Comment (lomigmegard) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala:150]: Use `StandardCharsets.US_ASCII` instead?

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 3]
```diff
@@ -81,19 +87,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
+    val firstIndexToEscape = valueBytes.indexWhere(isEscapingChar(_))
+    if (firstIndexToEscape == -1)
+      new String(valueBytes)
+    else
+      toAsciiStringSlow(valueBytes, firstIndexToEscape)
+  }
+
+  private def isEscapingChar(b: Byte): Boolean = b < ' ' || b >= '~' || b == '%'
+
+  private def toAsciiStringSlow(valueBytes: Array[Byte], riArg: Int): String = {
```
Comment (lomigmegard) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala:111]: move `toAsciiString`, `isEscapingChar` and `toAsciiStringSlow` to companion object? They seem to be all pure.

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala (hunk 1 above)
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala]: I know it seems like re-wrapping the same content but actually it's not

The following is not always true for any `value`: `new String(value.getBytes(StandardCharsets.UTF_8)) == value`. Consider `"a \ud801"`, its underlying array of bytes is `[97, 0, 32, 0, 1, -40]`. It's a UTF-16 string, note that even 'a' is encoded with 2 bytes. `"a \ud801".getBytes(StandardCharsets.UTF_8)` yields `[97, 32, 63]`, there are 2 effects: ascii characters are encoded with 1 byte and lone surrogate has been replaced with `?` (63 in ascii)

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 4]
```diff
@@ -81,19 +86,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
+        escapedBytes.update(wi + 1, HexArr((b >> 4) & 0xf))
+        escapedBytes.update(wi + 2, HexArr(b & 0xf))
+        wi += 3
+      } else {
+        escapedBytes.update(wi, b)
+        wi += 1
+      }
+      ri += 1
+    }
+    new String(jutil.Arrays.copyOf(escapedBytes, wi))
```
Comment (lomigmegard) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala:169]: Should be equivalent to the following, and being explicit for the charset is better instead of relying of the system's default:
`new String(escapedBytes, 0, wi, StandardCharsets.US_ASCII)`

**Review by patriknw — COMMENTED** [review]
thanks for contributing, I have a few questions related to performance...

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 5]
```diff
@@ -81,19 +85,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
-  override def parse(value: String): Try[`Status-Message`] = Try(new `Status-Message`(value))
+  override def parse(value: String): Try[`Status-Message`] = Try(new `Status-Message`(parseAsciiString(value)))
 
   def findIn(headers: immutable.Seq[HttpHeader]): Option[String] =
     headers.collectFirst { case h if h.is(name) => h.value() }
+
+  private val HexArr: Array[Byte] =
+    Array('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L626
```
[the remark this answers is not in the document]
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala]: Moved to `PercentEncoding.scala`

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 6]
```diff
@@ -81,19 +85,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
+  private def parseAsciiString(value: String): String = {
+    val transferEncoding = StandardCharsets.US_ASCII
+    // This can be made faster if necessary.
```
[the remark this answers is not in the document]
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala]: It's used only for errors

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala [hunk 7]
```diff
@@ -81,19 +85,86 @@ object `Status` extends ModeledCustomHeaderCompanion[`Status`] {
...
+    val source = value.getBytes(transferEncoding)
+    val buf = ByteBuffer.allocate(source.length)
```
[the remark this answers is not in the document]
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/headers.scala]: You're right, reworked according to the suggestion

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L582
+  object Encoder {
+    private val HexArr: Array[Byte] =
+      Array('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+    def encode(value: String): String = {
+      val valueBytes = value.getBytes(StandardCharsets.UTF_8)
+      val firstIndexToEscape = valueBytes.indexWhere(isEscapingChar(_))
+      if (firstIndexToEscape == -1)
+        new String(valueBytes)
```
Comment (viktorklang) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala:43]: Missing character encoding in String constructor here

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        value
+
+    private def decodeSlow(value: String): String = {
+      val source = value.getBytes(TransferEncoding)
+      val buf = ByteBuffer.allocate(source.length)
+      var i = 0
+      while (i < source.length) {
+        if (source(i) == '%' && i + 2 < source.length) {
+          try {
+            buf.put(Integer.parseInt(new String(source, i + 1, 2, TransferEncoding), 16).toByte)
```
Comment (viktorklang) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala:92]: Seems extremely costly to do this instead of looking at 2 bytes and create a number?

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala (hunk 8 above)
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala]: Corrected

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala (hunk 9 above)
Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala]: True, reworked according to the suggestion. I ran microbenchmark and the difference for non-contrived input was significant indeed - 40% of speedup

File: runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            i += 1
+          }
+        } else {
+          buf.put(source(i))
+          i += 1
+        }
+      }
+      new String(buf.array, 0, buf.position(), StandardCharsets.UTF_8)
+    }
+  }
```
Comment (viktorklang) [inline runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala:112]: It would be interesting to bench the solution above against the following:

```scala
private[this] final def decodeSlow(value: String): String = {
  @tailrec def loop(buf: ByteBuffer, source: Array[Byte], at: Int): String =
    if (at < source.length) {
      val cur = source(at)
      val skip = 
        if (cur != '%' || (at + 2) > source.length) {
          buf.put(cur)
          1
        } else {
          val ch0 = Character.digit(source(at + 1), 16)
          val ch1 = Character.digit(source(at + 2), 16)
          if (ch0 > -1 && ch1 > -1) {
            buf.put(((ch0 << 4) + ch1).toByte)
            3
          } else {
            buf.put(cur)
            1
          }
        }
      loop(buf, source, at + skip)
    } else new String(buf.array, 0, buf.position, StandardCharsets.UTF_8)

  val source = value.getBytes(TransferEncoding)
  val buf = ByteBuffer.allocate(value.length)

  loop(buf, source, 0)
}
```

Another possible improvement is to pass in the `idx` found by `indexOf('%')` and do an arraycopy from 0 … idx to avoid having to put those chars 1-by-1, as the worst-case the '%' is at the end of the string. Then you can start the `loop` at `idx` and progress from there.

Comment (patriknw) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala:112]: It's good that we try to find a decent performance level for this, but note that it's only for errors and that shouldn't be extremely performance sensitive.

Comment (note) [reply runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala:112]: With the code pasted by @viktorklang I got `0,652 ± 0,003  us/op` while with the code currently in the PR I got `0,609 ± 0,009  us/op`.

I agree with @patriknw that it's good to have reasonable performance but IMO it's not critical to get the best since it's only for errors

File: project/Common.scala [hunk 11]
```diff
@@ -51,6 +53,7 @@ object Common extends AutoPlugin {
       "-P:silencer:globalFilters=Use LazyList instead of Stream",
       // ignore imports in templates
       "-P:silencer:pathFilters=.*.txt"),
+    Compile / console / scalacOptions ~= (_.filterNot(consoleDisabledOptions.contains)),
```
Comment (raboof) [inline project/Common.scala:56]: this change doesn't really belong in this PR, but I like it anyway ;)

File: runtime/src/test/scala/akka/grpc/scaladsl/headers/HeadersSpec.scala [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      val examples = Table(
+        "examples",
+        "example%",
+        "yet another%%example",
+        "abc\ndef",
+        "abc\ndef\n",
+        "abc\r\ndef\r\n",
+        "abc\r\n\r\ndef\r\n",
+        "Καλημέρα κόσμε",
+        "コンニチハ",
```
Comment (raboof) [inline runtime/src/test/scala/akka/grpc/scaladsl/headers/HeadersSpec.scala:78]: 今日は!

**Review by raboof — APPROVED** [review]
Great, thanks!
